### ISSUE TITLE
Compile CuPy kernels before import

### ISSUE THREAD

**Issue by mnicely** [cupy/cupy#3314]
Would it be possible to compile multiple/all CuPy raw kernels in a build-like process? The idea being that the kernels would be compiled during/before package install. That way they are compiled before import. @awthomp

**Comment by leofang** [comment]
If you have all of your CUDA kernels written in .cu files, then you can write a few Python functions that are hooked up to `setup.py`, in which you compile them to .cubin files using nvcc. Then, they can be loaded at runtime using `RawModule`. In one of our projects we have a few snippets adapted from https://github.com/cupy/cupy/blob/master/install/build.py that enables this build process.

**Comment by leofang** [comment]
A common pitfall: You need to remember to include the cubin files when distributing your package 😂 Add a line like `include ./*.cubin` to `MANIFEST.in` should suffice.

**Comment by jakirkham** [comment]
How big do these get? Should we consider doing this with the conda-forge `cupy` package?

**Comment by mnicely** [comment]
@jakirkham I need another week or two before I can start testing. I can tell you then!

**Comment by jakirkham** [comment]
Ah sorry for the confusion that was a question for @leofang 🙂

**Comment by leofang** [comment]
Sorry, John, for missing your ping. I don't get your question 😅

**Comment by jakirkham** [comment]
Sorry, how big do the `*.cubin` files get?

**Comment by leofang** [comment]
Hi John, I still don't understand. Did you mean you wanna bundle cuSignal's cubin files to the conda-forge `cupy` package? Why?

**Comment by jakirkham** [comment]
No, but we might want to bundle CuPy's `*.cubin` files in the `cupy` package or cuSignal might want to bundle `*.cubin` in a `cusignal` package. The benefit would be these are ready to go and CuPy could just pick up the `*.cubin` files and use them. If we started doing this, it would be good to know how much larger the package would be. Does that make more sense?

**Comment by leofang** [comment]
Ahhh, so this was what @mnicely meant?! I completely misunderstood and thought he simply wanna distribute the cubins for his custom RawKernels in a cuSignal release 😂

I think there would be many cubins to be generated. I don't have good suggestion. File size might not be the main concern, but the number of files. There was a time I tried `rm -rf ~/.cupy/kernel_cache/` and it failed, because the number of files exceeded max inode (or something filesystem-related). 

Maybe @kmaehashi @emcastillo have thought about it?

**Comment by jakirkham** [comment]
That's my understanding. Well that's part of it. Being able to do this for cuSignal kernels and include them in a `cusignal` package would also be useful. Feel free to hop in @mnicely if I'm misinterpreting things here 🙂

Good point! Do you recall how many? Is it possible we are being too conservative when generating a unique name for each `*.cubin` file?

Would certainly be curious to hear others thoughts as well?

**Comment by mnicely** [comment]
Sorry for the late response. 

Originally, I was wondering if it would be possible to precompile all the CuPy raw kernels, in cuSignal, before user even installs cuSignal. Right now we give the users the ability to precompile all CuPy kernels before first call. If the kernel is not compiled, it would be at runtime. 

But it would be nice if we can remove the compile process all together. @leofang pointed me to some sample code but I've been unable to check it out yet. I think we currently have 8 raw kernels (more coming). Another benefit for pushing cuSignal with precompiled kernels it that we can compile all possible data types for each kernels, saving even more time.

All this gets a little muddy when we start talking about added the templates @leofang and I have been working on but we can visit that later.

That being said, it is my understanding that I need to remove all the CuPy raw kernels from our `.py` files and put them in `.cu` files and create `.cubin` files.

@jakirkham Did that help or make things more confusing??

Adding @awthomp in case I forgot something.

**Comment by mnicely** [comment]
@jakirkham @leofang 
Did I completely misunderstand your questions? This might make for sense to me once I start adapting the code per @leofang's suggestions.

**Comment by leofang** [comment]
Frankly I was a bit lost. But let me try to summarize what's discussed above in my understanding:
1. precompile kernels in .cu: Just hook up the nvcc compilation in `setup.py` as suggested in https://github.com/cupy/cupy/issues/3314#issuecomment-621967237.
2. precompile `RawKernel`s: currently CuPy doesn't provide API to do that. I suggested to put the implementation in .cu and apply 1 above. This has an additional benefit that the .cu files can be read in Python to instantiate `RawKernel`s (so for maintenance it's just a matter of copy-and-paste without refactoring the code).
3. precompile CuPy built-in functions (ex: ufuncs and `_SimpleReductionKernel`s): I have no idea.

**Comment by jakirkham** [comment]
Sorry for the slow reply @mnicely. Yeah it makes sense.

The first part is what I thought you were after with this issue. Didn't realize you wanted to hook in compiled CUDA code into CuPy as well. Though that makes sense if you are wanting to hand tune these.

Yeah templating will just mean that you need to specify every configuration of a template you want to precompile. This is the tradeoff when trying to solve this problem at build time instead of at runtime.

Initially thought we were just trying to solve a packaging issue. However now it sounds like we are performance chasing, is that right? If so, it would be good to have a new issue about the performance issues you are running into and what you are hoping to see improved.

**Comment by mnicely** [comment]
> However now it sounds like we are performance chasing, is that right?

At this point I believe so. @leofang's #1 suggestion should solve the functionality aspect of my issue. Once I get this integrate I can dive further into performance aspect. I'll open a new issue if needed.

**Comment by jakirkham** [comment]
FWIW part of my reason for asking about performance is to see if there is low hang fruit if we move along that axis. It seems there is a fair bit to do on the external kernel path. So if performance fixes can be made without those changes, maybe we can solve the underlying issue faster. 😉 Defer to you on which direction is most appropriate (as you know more about the problems you are encountering that I do 🙂).

**Comment by leofang** [comment]
In the way things are written today, it is impossible to fix.
